### DISCUSSION TITLE
Callback function after entire array has been processed in the dom.

### DISCUSSION THREAD

**Discussion by Shashank-Bhatt1** [alpinejs/alpine#3641] [1. Help]
Hi I am trying to update slick slider which is bounded to some array to render elements inside it but not able to figure out the way to update it via it's refresh method only after array has been processed in the dom.

I tried to take another array which returns the spliced version of original array when looked up for it and it is working quite nicely but when I tried to update slider via refresh method it just always shows first three elements only and not the one which are sorted by distances.

```
Alpine.data('shops', () => ({
    shops: [],
    brandshops: [],
    get firstThreeBrandShops() {
        return this.brandshops.slice(0, 3);
    },
    init() {
        this.$watch('firstThreeBrandShops', (value, oldValue) => {
            $('#shops-section-slider.slick-initialized').slick('refresh');
        })
    },
    trigger: {
        ['@nearbyshopsupdated.window'](event) {
            let distances = event.detail.distances;
            distances.forEach((distance, index) => {
                let tempShop = this.shops.find(shop => shop.id === distance.storeid);

                tempShop.distanceText = distance.distanceText;
                tempShop.distanceVal = distance.distanceVal;
            })

            this.brandshops.sort((first, second) => {
                return first.distanceVal - second.distanceVal;
            });

        }
    }
}
))
```

and in template part

```
<div class="row shops-section-wrapper" id="shops-section-slider">
    <template x-for="(shop,index) in firstThreeBrandShops" :key="shop.id">
        <div>...Detail of shops set via slider</div>
    </template>
</div>

``` 

I need a way (callback or async type) to only refresh the slider when all of the elements have been sorted and even the dom is also updated with this new `firstThreeBrandShops` array. 

If i don't refresh the slider then it's showing correct result somehow but refreshing the slider is needed after updating array because then the slider items are not rendering in correct width.

 I have also tried using $nextTick in the init with same result.
```
 let self = this;
  this.$watch('firstThreeBrandShops', (value, oldValue) => {
 self.$nextTick(() => {
        $('#shops-section-slider.slick-initialized').slick('refresh');
 });
 })
```

**Comment by ekwoka** [comment]
1. don't use jquery. This isn't 2015

2. the nextTick should 100% work.

```html
<div class="row shops-section-wrapper" id="shops-section-slider" 
  x-init="$watch('brandShops', () => $nextTick(()=> slider.refresh()))">
    <template x-for="(shop,index) in firstThreeBrandShops" :key="shop.id">
        <div>...Detail of shops set via slider</div>
    </template>
</div>
```

If you watch the actual brand shops, you'll better be able to respond to changes.

**Reply by Shashank-Bhatt1** [reply]
Hi @ekwoka I have found that the issue was within the slider itself rather than alpine but at present i didn't have alpine slider component so for slick slider purpose only, I have used jQuery (e.g. init and refresh). 

It turned out that slider itself stores data somewhere and when we update array and refresh slick instance afterwards, alpine first updates the template node thus updating the items in html dom, however when we call refresh of slick instance, it just set to the previous slides which was there before array update and so even after updating array I kept getting the older array. 

I ended up doing something like this which is at least working for now. It would be great if you can provide any minimal demo with updating array and still maintaining slider and it's correct width.

```
self.$watch('firstThreeBrandShops',(currentShops,oldShops) => {
                $('#shops-section-slider.slick-initialized').slick('slickRemove', null, null, true);
                currentShops.forEach(function(shop,index) {
                    $('#shops-section-slider.slick-initialized').slick('slickAdd', self.getSliderTmpl(shop));
                })
                $('#shops-section-slider.slick-initialized').slick('refresh');
            }) 
```
and returning html from getSliderTmpl.
```
 getSliderTmpl(shop) {
    return `<div class="col-12 col-md-6 col-lg-4 col-xl-4 d-flex flex-column shops-section shops-section-js">
        <div class="row flex-grow-1 flex-column flex-xxl-row">
            <div class="col-12 col-xxl-6">
                <div class="shops-img position-relative">
                    <strong class="shops-title-card d-block d-xxl-none">${shop.title}</strong>
                </div>
            </div>
            
        </div>
    </div>`;
}
```

**Reply by Shashank-Bhatt1** [reply]
I have also noted that `<template>` node which renders the for loop is also taken into account as slide and there is no way around this thing. There is setting in slick to designate slide via it's slide option to particular class but this creates more issues by putting template completely outside of slick structure.
